### PR TITLE
Mod_rewrite refactor and cleanup

### PR DESCRIPTION
**Pull request by ZestyBread**
Moving around a ton of stuff and cleans up mod_rewrite to flow functionally like UrlRewrite. 
@davidfowl @Tratcher @muratg 

This PR is a lot of changes, without a lot of substance. Going to schedule a code review for this.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/FileParser.cs [hunk 1]
```diff
@@ -33,58 +31,55 @@ public static List<Rule> Parse(TextReader input)
...
-                    //    ModRewriteBase.Base = tokens[1];
-                    //}
-                    //else
-                    //{
-                    //    throw new FormatException("");
-                    //}
-                    //break;
+                        throw new NotImplementedException("RewriteBase is not implemented");
                     case "RewriteCond":
                         {
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/FileParser.cs]: style: curly brackets are not required in case statements.

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/FileParser.cs [hunk 2]
```diff
@@ -33,58 +31,55 @@ public static List<Rule> Parse(TextReader input)
...
+                                    flags = FlagParser.Parse(tokens[3]);
+                                }
+
+                                builder.AddConditionFromParts(pattern, condActionParsed, flags);
                             }
-                            else
+                            catch (FormatException formatException)
                             {
-                                throw new FormatException();
+                                throw new FormatException(Resources.FormatError_ModRewriteParseError(formatException.Message, lineNum));
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/FileParser.cs]: include inner exception?

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRewriteAction.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+        public override RuleResult ApplyAction(HttpContext context, MatchResults ruleMatch, MatchResults condMatch)
+        {
+            var pattern = Url.Evaluate(context, ruleMatch, condMatch);
+
+            // TODO PERF, substrings, object creation, etc.
+            if (pattern.IndexOf("://") >= 0)
+            {
+                string scheme = null;
+                var host = new HostString("");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRewriteAction.cs]: It's odd to initialize out parameters. `HostString host;` should be adequate.

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRewriteAction.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+                var path = PathString.Empty;
+                var query = QueryString.Empty;
+                var fragment = FragmentString.Empty;
+                UriHelper.FromAbsolute(pattern, out scheme, out host, out path, out query, out fragment);
+
+                if (query.HasValue)
+                {
+                    if (QueryStringAppend)
+                    {
+                        context.Request.QueryString = query.Add(context.Request.QueryString);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRewriteAction.cs]: reverse this so you append the new query to the end of the existing one.

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRewriteAction.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            }
+            else
+            {
+                var split = pattern.IndexOf('?');
+                if (split >= 0)
+                {
+                    var path = pattern.Substring(0, split);
+                    if (path.StartsWith(ForwardSlash))
+                    {
+                        context.Request.Path = new PathString(path);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRewriteAction.cs]: `PathString.FromUriComponent(path)`

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ParsedModRewriteCondition.cs [hunk 6]
```diff
@@ -3,20 +3,21 @@
...
-        public ConditionType Type { get; set; }
-        public OperationType Operation { get; set; }
+        public ConditionType ConditionType { get; set; }
+        public OperationType OperationType { get; set; }
         public string Operand { get; set; }
-        public ParsedModRewriteExpression(bool invert, ConditionType type, OperationType operation, string operand)
+
+        public ParsedModRewriteInput() { }
+
+        public ParsedModRewriteInput(bool invert, ConditionType type, OperationType operation, string operand)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ParsedModRewriteCondition.cs]: rename the parameters to match?

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/FileParser.cs [hunk 7]
```diff
@@ -33,71 +31,64 @@ public static List<Rule> Parse(TextReader input)
...
+                                flags = FlagParser.Parse(tokens[3]);
                             }
-                            conditions.Add(builder.Build());
-                            break;
+
+                            builder.AddConditionFromParts(pattern, condActionParsed, flags);
                         }
+                        catch (FormatException formatException)
+                        {
+                            throw new FormatException(Resources.FormatError_ModRewriteParseError(formatException.Message, lineNum), formatException);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/FileParser.cs]: No need to duplicate the Message from the inner exception

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRewriteAction.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+                if (split >= 0)
+                {
+                    var path = pattern.Substring(0, split);
+                    if (path.StartsWith(ForwardSlash))
+                    {
+                        context.Request.Path = PathString.FromUriComponent(path);
+                    }
+                    else
+                    {
+                        context.Request.Path = new PathString(ForwardSlash + path);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRewriteAction.cs]: FromUriComponent

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRewriteAction.cs]: + 2 more instances in this file

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewriteRule.cs [hunk 9]
```diff
@@ -1,208 +1,54 @@
...
 
-using System;
 using System.Collections.Generic;
-using Microsoft.Net.Http.Headers;
-using System.Text.RegularExpressions;
-using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Rewrite.Internal.Shared;
 
-namespace Microsoft.AspNetCore.Rewrite.Internal.ModRewrite
+namespace Microsoft.AspNetCore.Rewrite.Internal
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewriteRule.cs]: Why did the namespace change?

File: src/Microsoft.AspNetCore.Rewrite/Internal/Shared/Pattern.cs [hunk 10]
```diff
@@ -5,10 +5,12 @@
 using System.Text;
 using Microsoft.AspNetCore.Http;
 
-namespace Microsoft.AspNetCore.Rewrite.Internal.UrlRewrite
+namespace Microsoft.AspNetCore.Rewrite.Internal.Shared
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/Shared/Pattern.cs]: Shared items can stay in the parent Internal namespace, or put them in a sub-namespace that describes their function.

File: src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PatternSegments/DateTimeSegment.cs [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+                case "TIME_SEC":
+                    _portion = DateTimePortion.Day;
+                    break;
+                case "TIME_WDAY":
+                    _portion = DateTimePortion.Day;
+                    break;
+                case "TIME":
+                    _portion = DateTimePortion.Day;
+                    break;
+            }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PatternSegments/DateTimeSegment.cs]: default? throw notimplentedexception(segment)

File: src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PatternSegments/IsHttp2Segment.cs [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Rewrite.Internal.Shared.PatternSegments
+{
+    public class IsHttp2Segment : PatternSegment
+    {
+        public override string Evaluate(HttpContext context, MatchResults ruleMatch, MatchResults condMatch)
+        {
+            return context.Request.Scheme == "http2" ? "on" : "off";
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PatternSegments/IsHttp2Segment.cs]: This is not how to detect http/2. We don't surface this information anywhere at the moment so you'll have to disable this feature.

File: src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PatternSegments/IsIPV6Segment.cs [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Rewrite.Internal.Shared.PatternSegments
+{
+
+    public class IsIPV6Segment : PatternSegment
+    {
+        public override string Evaluate(HttpContext context, MatchResults ruleMatch, MatchResults condMatch)
+        {
+            return context.Connection.LocalIpAddress.AddressFamily == AddressFamily.InterNetworkV6 ? "on" : "off";
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PatternSegments/IsIPV6Segment.cs]: Use RemoteIpAddress instead. That will account for x-forwarded-for and reverse proxy scenarios. Also, the IP may be null

File: src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PreActions/ChangeCookiePreAction.cs [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    public class ChangeCookiePreAction : PreAction
+    {
+        public ChangeCookiePreAction(string cookie)
+        {
+            // TODO
+        }
+
+        public override void ApplyAction(HttpContext context, MatchResults ruleMatch, MatchResults condMatch)
+        {
+            // modify the cookies
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PreActions/ChangeCookiePreAction.cs]: throw notimplmented?

File: src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PreActions/ChangeEnvironmentPreAction.cs [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+    {
+
+        public ChangeEnvironmentPreAction(string env)
+        {
+            // TODO
+        }
+
+        public override void ApplyAction(HttpContext context, MatchResults ruleMatch, MatchResults condMatch)
+        {
+            // Do stuff to modify the env
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PreActions/ChangeEnvironmentPreAction.cs]: throw notimplmented?

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewriteRule.cs (hunk 9 above)
Comment (ZestyBread) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewriteRule.cs]: It reflects the actual file structure.

File: src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PatternSegments/DateTimeSegment.cs (hunk 11 above)
Comment (ZestyBread) [inline src/Microsoft.AspNetCore.Rewrite/Internal/Shared/PatternSegments/DateTimeSegment.cs]: More appropriately formatexception- unrecognized time specifier.

File: src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRedirectAction.cs [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace Microsoft.AspNetCore.Rewrite.Internal.ModRewrite
+{
+    public class ModRewriteRedirectAction : UrlAction
+    {
+        public int StatusCode { get; }
+        public bool QueryStringAppend { get; }
+        public bool QueryStringDelete { get; }
+        public bool EscapeBackReferences { get; }
+        public ModRewriteRedirectAction(
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/ModRewrite/ModRewriteRedirectAction.cs]: nit: new line

File: src/Microsoft.AspNetCore.Rewrite/Internal/PatternSegments/ServerProtocolSegment.cs [hunk 17]
```diff
@@ -0,0 +1,17 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+
+namespace Microsoft.AspNetCore.Rewrite.Internal.PatternSegments
+{
+
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/PatternSegments/ServerProtocolSegment.cs]: nit: extra line

File: src/Microsoft.AspNetCore.Rewrite/Internal/PreActions/ChangeEnvironmentPreAction.cs [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.Rewrite.Internal.PreActions
+{
+    public class ChangeEnvironmentPreAction : PreAction
+    {
+
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/PreActions/ChangeEnvironmentPreAction.cs]: nit: extra line

File: src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RewriteAction.cs [hunk 19]
```diff
@@ -19,28 +19,28 @@ public RewriteAction(RuleTerminiation result, Pattern pattern, bool clearQuery)
...
+                FragmentString fragment;
                 UriHelper.FromAbsolute(pattern, out scheme, out host, out path, out query, out fragment);
 
-                context.Request.Scheme = scheme;
-                context.Request.Host = host;
-                context.Request.Path = path;
-                context.Request.QueryString = query.Add(context.Request.QueryString);
+                context.HttpContext.Request.Scheme = scheme;
+                context.HttpContext.Request.Host = host;
+                context.HttpContext.Request.Path = path;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RewriteAction.cs]: What about PathBase?

**Comment by Tratcher** [comment]
:shipit:

File: src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RewriteAction.cs (hunk 19 above)
Comment (ZestyBread) [inline src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/RewriteAction.cs]: https://github.com/aspnet/BasicMiddleware/issues/59, filed and will be handled in another pr.
